File: frontend/src/utilities/selectionSort.js
const selectionSort = async (
  array,
  updateArray,
  highlightIndices,
  delay,
  updateSkip,
  setSwapCount,
  setSortTime,
  setArrayAccesses,
  setComparisonCount
) => {
  const start = performance.now();
  let swapCount = 0; // To track the swaps
  let comparisonCount = 0; // To track the comparisons
  let arrayAccess = 0; // To track the array accesses

  const tempArray = [...array]; // Create a copy to avoid mutating the original array
  let updateCounter = 0; // Counter to control throttling

  // Traverse through all elements
  for (let i = 0; i < tempArray.length; i++) {
    let minIndex = i;  // Assume the current index as the minimum

    // Traverse the remaining unsorted portion of the array
    for (let j = i + 1; j < tempArray.length; j++) {
      highlightIndices([minIndex, j]);  // Highlight the current pair of elements

      comparisonCount++; // Track each comparison
      setComparisonCount(comparisonCount);

      // Increment array accesses for both elements in comparison
      arrayAccess += 2; 
      setArrayAccesses(arrayAccess);

      // Update minIndex if a smaller element is found
      if (tempArray[j] < tempArray[minIndex]) {
        minIndex = j;
      }

      // Increment the update counter and throttle updates
      updateCounter++;
      if (updateCounter % updateSkip === 0) {
        updateArray([...tempArray]);  // Update the array visually
        await new Promise(resolve => setTimeout(resolve, delay));  // Add delay
      }
    }

    // Swap elements if necessary
    if (minIndex !== i) {
      swapCount++; // Increment swap count
      setSwapCount(swapCount);
      [tempArray[i], tempArray[minIndex]] = [tempArray[minIndex], tempArray[i]];
      arrayAccess += 4; // Two accesses for the elements being swapped
      setArrayAccesses(arrayAccess);
    }

    // Final update after each pass
    updateArray([...tempArray]);
    await new Promise(resolve => setTimeout(resolve, delay));  // Add delay after swap
  }

  highlightIndices([-1, -1]);  // Clear the highlights after sorting is complete
  setSortTime(performance.now() - start); // Set the total time taken for sorting
};

export default selectionSort;

File: backend/api/utils/selection_sort.py
# utils/sorting.py
import asyncio

# Async Selection Sort Algorithm
async def selection_sort(arr, send_fn, delay, throttle):
    comparisons = 0
    arrayAccesses = 0
    swaps = 0
    highlighted_indices = []
    n = len(arr)
    
    for i in range(n):
        min_index = i
        for j in range(i + 1, n):
            highlighted_indices = [min_index, j]  # Highlight current min and j
            comparisons += 1
            arrayAccesses += 2  # Reading arr[min_index] and arr[j]
            
            if arr[j] < arr[min_index]:  # Find the smallest element in unsorted part
                min_index = j
            arrayAccesses += 2  # Access the element for comparison
            
            # Send the data after each iteration (allow visualization update)
            if i % throttle == 0:
                await send_fn(arr, comparisons, arrayAccesses, swaps, highlighted_indices)
                await asyncio.sleep(delay)  # Delay for visualization (non-blocking)
        
        # Swap the found minimum element with the first element
        if min_index != i:
            arr[i], arr[min_index] = arr[min_index], arr[i]
            swaps += 2  # Swapping two elements
            arrayAccesses += 4  # Accessing both elements for swap
        
        # Send the data after each iteration (allow visualization update)
        if i % throttle == 0:
            await send_fn(arr, comparisons, arrayAccesses, swaps, highlighted_indices)
            await asyncio.sleep(delay)  # Delay for visualization (non-blocking)
        
    return arr, comparisons, arrayAccesses, swaps
